src/Plugin/RulesAction/SystemPageRedirect.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+/**
+ * @file
+ * Contains Drupal\rules\Plugin\Action\SystemPageRedirect.
+ */
+
+namespace Drupal\rules\Plugin\RulesAction;
+
+use Drupal\Core\Logger\LoggerChannelFactoryInterface;
+use Drupal\Core\Plugin\ContainerFactoryPluginInterface;
+use Drupal\rules\Core\RulesActionBase;
+use Psr\Log\LogLevel;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * Provides "Page redirect" rules action.
+ *
+ * @RulesAction(
+ *   id = "rules_page_redirect",
+ *   label = @Translation("Page redirect"),
+ *   category = @Translation("System"),
+ *   context = {
+ *     "url" = @ContextDefinition("string",
+ *       label = @Translation("URL"),
+ *       description = @Translation("A Drupal path, path alias, or external URL to redirect to. Enter (optional) queries after ? and (optional) anchor after #."),
+ *     ),
+  *   }
+ * )
+ *
+ */
+class SystemPageRedirect extends RulesActionBase implements ContainerFactoryPluginInterface {
+
+  /**
+   * The logger for the rules channel.
+   *
+   * @var \Psr\Log\LoggerInterface
+   */
+  protected $logger;
+
+  /**
+   * The current request.
+   *
+   * @var null|\Symfony\Component\HttpFoundation\Request
+   */
+  protected $request;
+
+  /**
+   * Constructs a PageRedirect object.
+   *
+   * @param array $configuration
+   *   A configuration array containing information about the plugin instance.
+   * @param string $plugin_id
+   *   The plugin ID for the plugin instance.
+   * @param mixed $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\Core\Logger\LoggerChannelFactoryInterface $loggerFactory
+   *   The logger factory service.
+   * @param \Symfony\Component\HttpFoundation\RequestStack $requestStack
+   *   The request stack service.
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, LoggerChannelFactoryInterface $loggerFactory, RequestStack $requestStack) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->logger = $loggerFactory->get('rules');
+    $this->request = $requestStack->getCurrentRequest();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('logger.factory'),
+      $container->get('request_stack')
+    );
+  }
+
+  /**
+   * Redirect to a provided url at the end of the request.
+   *
+   * @param string $url
+   *   Redirect destination url.
+   */
+  protected function doExecute($url) {
+    $current_path = $this->request->getPathInfo();
```

Comment:
I think we need to inject the path.current service and use ->getPath() to get the current path, Otherwise we might get an alias here, but we want the unaliased path.